src/Sound/Tidal/Stream/Process.hs

```diff
@@ -124,27 +124,29 @@ doTick stateMV busMV playMV globalFMV cxs listen (st,end) nudge ops =
                 hPutStrLn stderr $ "Failed to send. Is the '" ++ oName target ++ "' target running? " ++ show e
       putMVar stateMV sMap'')
 
-processCps :: Clock.LinkOperations -> [Event ValueMap] -> IO [ProcessedEvent]
-processCps ops = mapM processEvent
+processCps :: Clock.ClockConfig -> Clock.ClockRef -> Link.SessionState -> [Event ValueMap] -> IO [ProcessedEvent]
+processCps cconf cref ss = mapM processEvent
   where
     processEvent ::  Event ValueMap  -> IO ProcessedEvent
     processEvent e = do
       let wope = wholeOrPart e
           partStartCycle = start $ part e
-          partStartBeat = (Clock.cyclesToBeat ops) (realToFrac partStartCycle)
+          partStartBeat = (Clock.cyclesToBeat cconf) (realToFrac partStartCycle)
           onCycle = start wope
-          onBeat = (Clock.cyclesToBeat ops) (realToFrac onCycle)
+          onBeat = (Clock.cyclesToBeat cconf) (realToFrac onCycle)
           offCycle = stop wope
-          offBeat = (Clock.cyclesToBeat ops) (realToFrac offCycle)
-      on <- (Clock.timeAtBeat ops) onBeat
-      onPart <- (Clock.timeAtBeat ops) partStartBeat
+          offBeat = (Clock.cyclesToBeat cconf) (realToFrac offCycle)
+      on <- Clock.timeAtBeat cconf ss onBeat
+      onPart <- Clock.timeAtBeat cconf ss partStartBeat
       when (eventHasOnset e) (do
         let cps' = Map.lookup "cps" (value e) >>= getF
-        maybe (return ()) (\newCps -> (Clock.setTempo ops) ((Clock.cyclesToBeat ops) (newCps * 60)) on) $ coerce cps'
+        maybe (return ()) (\newCps -> Clock.setCPS cconf cref newCps) (fmap toRational cps')
```

Comment:
I expect this to behave differently? Could cause subtle differences when events change cps and the events coming afterwards won't use the new cps until next tick.

Reply:
yes it might be a difference of a tick, the reason i did here like this is to avoid carrying around an additional sessionstate that would keep track of time changes (this was previously possible, since the LinkOperations would carry this specific version of setTempo)
i'm not so sure if the complication is worth it here

Reply:
I'm not really active in this project at the moment so someone else need to make the call of whether the complexity is worth it. But now you see the motivation for the current design.